File: java/testsuite/Jogl4Examples/PolyhedralBoundedSolidExample/src/animation/SolidAnimationController.java
package animation;

import com.jogamp.opengl.awt.GLCanvas;

import models.DebuggerModel;
import vsdk.toolkit.animation.AnimationEvent;
import vsdk.toolkit.animation.AnimationEventGenerator;
import vsdk.toolkit.animation.AnimationListener;

public class SolidAnimationController
{
    private static final double ROTATION_STEP_DEGREES = 1.0;

    private AnimationEventGenerator animator;
    private boolean started;

    public void start(DebuggerModel model, GLCanvas panel)
    {
        if ( started ) {
            return;
        }

        animator = new AnimationEventGenerator();
        animator.addAnimationListener(new AnimationListener() {
            @Override
            public void tick(AnimationEvent e)
            {
                if ( !model.isSolidAnimationEnabled() ) {
                    return;
                }
                model.rotateSolidAroundZDegrees(ROTATION_STEP_DEGREES);
                panel.repaint();
            }
        });

        Thread t = new Thread(animator);
        t.setDaemon(true);
        t.start();
        started = true;
    }

    public void toggleAnimation(DebuggerModel model)
    {
        model.toggleSolidAnimationEnabled();
    }
}

File: typescript/testsuite/VitralWebTestsuiteContainer/src/WebGLExamples/PolyhedralBoundedSolidExample/animation/solid-animation-controller.ts
import { AnimationEventGenerator, AnimationListener, type AnimationEvent } from '@vitral/base';
import type { DebuggerModel } from '../models/debugger-model';

/**
 * Port of
 * `java/testsuite/Jogl4Examples/PolyhedralBoundedSolidExample/src/animation/SolidAnimationController.java`.
 *
 * One `AnimationEventGenerator` whose listener, on every tick while the
 * model's solid animation is on, turns the solid one degree about Z and asks
 * the canvas to repaint. Java hands the generator to a daemon `Thread`; the
 * ported generator's `run()` arms the host timer instead, as its own record
 * explains, which is why this class also has a `stop`, called when the module
 * is torn down, where Java's thread simply dies with the process.
 *
 * Java's `panel.repaint()` is the component's repaint request, passed in.
 */
export class SolidAnimationController {
  private static readonly ROTATION_STEP_DEGREES = 1.0;

  private animator: AnimationEventGenerator | null = null;
  private started = false;

  start(model: DebuggerModel, repaint: () => void): void {
    if (this.started) {
      return;
    }

    this.animator = new AnimationEventGenerator();
    this.animator.addAnimationListener(
      new (class extends AnimationListener {
        override tick(_e: AnimationEvent): void {
          if (!model.isSolidAnimationEnabled()) {
            return;
          }
          model.rotateSolidAroundZDegrees(SolidAnimationController.ROTATION_STEP_DEGREES);
          repaint();
        }
      })(),
    );

    this.animator.run();
    this.started = true;
  }

  stop(): void {
    this.animator?.stop();
    this.animator = null;
    this.started = false;
  }

  toggleAnimation(model: DebuggerModel): void {
    model.toggleSolidAnimationEnabled();
  }
}
